### ISSUE TITLE
ducc vs libsharp SHT performance scales like lmax

### ISSUE THREAD

**Issue by xzackli** [simonsobs/pixell#178]
@mreineck has a new package called [ducc](https://gitlab.mpcdf.mpg.de/mtr/ducc) that performs, among other things, SHTs. It scales like lmax compared to libsharp used by pixell. On a typical della node,
```
pip3 install ducc0

salloc --nodes=1 --ntasks-per-node=28 --time=1:00:00
OMP_NUM_THREADS=28 ipython
```
Suppose we choose a reduced band limit, `lmax = 10800` for typical ACT pixels.
```python
from pixell import enmap,utils, curvedsky,powspec
import os
import numpy as np
import ducc0
nthreads=int(os.environ.get('OMP_NUM_THREADS'))  # for fairness

lmax = 10800
mmax = lmax
ps = (np.arange(lmax+1) + 1.0)**(-2)  # ℓ⁻² spectrum
shape,wcs = enmap.fullsky_geometry(res=np.deg2rad(0.5/60.))
imap = curvedsky.rand_map(shape,wcs,ps=ps,spin=0)
imap_reshaped = np.array(imap).reshape((1,) + imap.shape)
%timeit alm = curvedsky.map2alm(imap, lmax=lmax)
%timeit alm2 = ducc0.sht.experimental.analysis_2d(map=imap_reshaped, lmax=lmax, mmax=mmax, spin=0, geometry="CC", nthreads=nthreads)
```
```
13 s ± 46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
7.07 s ± 26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
ducc seems to be twice as fast here. However if we choose the maximum `lmax = 21600` (i.e. near Nyquist),
```python
lmax = 21600
mmax = lmax
ps = (np.arange(lmax+1) + 1.0)**(-2)  # ℓ⁻² spectrum
shape,wcs = enmap.fullsky_geometry(res=np.deg2rad(0.5/60.))
imap = curvedsky.rand_map(shape,wcs,ps=ps,spin=0)
imap_reshaped = np.array(imap).reshape((1,) + imap.shape)
%timeit alm = curvedsky.map2alm(imap, lmax=lmax)
%timeit alm2 = ducc0.sht.experimental.analysis_2d(map=imap_reshaped, lmax=lmax, mmax=mmax, spin=0, geometry="CC", nthreads=nthreads)
```
```
41.7 s ± 134 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
42.7 s ± 191 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by mreineck** [comment]
Hi @xzackli, great to hear that this version is more efficient for you  at least in some circumstances!

You may get even better performance if you install via
```
pip3 install --no-binary ducc0 ducc0
```
since that will compile specifically for the detected CPU. In the best case, this could give you another factor of 2.

I don't know why there are no improvements for the `lmax = 21600` case. I'll try to run a few benchmarks myself; so far I haven't tested too much at such high band widths.

**Comment by mreineck** [comment]
P.S.: One thing you could try is to set the number of threads to the number of physical cores on the node, not the number of virtual ones (which is the typical value of OMP_NUM_THREADS). I imagine this could help at high band limits, since there may be less cache thrashing.

**Comment by xzackli** [comment]
Wow! Indeed, compiling for the detected CPU seems to lead to another factor of 2.

For `lmax = 10800`,
```
13.1 s ± 75.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)   # pixell using libsharp2
3.74 s ± 13.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)   # ducc
```
For `lmax = 21600`
```
41.4 s ± 83.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
17.5 s ± 99.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
We should certainly switch to `ducc` when it becomes less experimental.

**Comment by mreineck** [comment]
Sorry, the `experimental` is somewhat misleading here ... the algorithms are actually fairly mature; what's experimental is the interface presented by the functions in `ducc0.sht.experimental`. I wanted to have the chance to still adjust details of the interface based on user feedback, and that's why I didn't directly putt them into `ducc0.sht`.

So I would certainly be happy hear about your experiences with the code!

**Comment by xzackli** [comment]
It wasn't clear to me from the docs what the difference was between the analysis functions and their counterparts named `_2d`. Is it a matter of data memory layout?

**Comment by mreineck** [comment]
That is part of it. The `_2d` functions only work for maps with rings that all have the same number of pixels and start at the same azimuth (so they can|t be used for Healpix, for example), and they only support a subset of possible `theta` distributions. That makes calling them much more intuitive than than the fully general `alm2map` function. And only the `_2d` functions have a (more or less) well-defined analysis function.

Part of the speedup in `ducc` comes from the fact that the CC grid can actually be reduced to half the number of rings that you are currently using for maps with a given `lmax`. The idea behind this is sketched in https://www.mpa-garching.mpg.de/~martin/shtnote.pdf.

**Comment by mreineck** [comment]
> Part of the speedup in ducc comes from the fact that the CC grid can actually be reduced to half the number of rings that you are currently using for maps with a given lmax. The idea behind this is sketched in https://www.mpa-garching.mpg.de/~martin/shtnote.pdf.

This unforunately also implies that this nice speedup is lost if you transform maps that only cover parts of the theta range instead of the full one ...

**Comment by xzackli** [comment]
Oh, that's very pertinent! As you might know, ACT/SO maps are usually not full sky. Pixell by default will zero-pad the map so the rings are complete, and then pass those rings to libsharp. Such a map is complete in phi (full rings) but incomplete in theta -- are you saying that libsharp and ducc would perform similarly?

I should benchmark some partial-sky SHTs.

**Comment by mreineck** [comment]
For partial maps, `ducc` won't be significantly faster than `libsharp` (unless you compiled libsjarp without native CPU optimization), but it should not be slower either. In the longer term I certainly suggest to switch.

**Comment by msyriac** [comment]
Implemented in #216
